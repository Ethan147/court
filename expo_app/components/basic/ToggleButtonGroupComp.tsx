import React, { useState } from "react";
import {
  View,
  Text,
  Pressable,
  StyleProp,
  TextStyle,
  ViewStyle,
} from "react-native";

interface ToggleButtonGroupCompProps {
  onValueChange?: (value: string) => void;
  buttons: string[];
  label: string;
  passStyles?: {
    toggleButtonGroupCompContainer?: StyleProp<ViewStyle>;
    toggleButtonGroupCompLabel?: StyleProp<TextStyle>;
    toggleButtonGroupCompButtons?: StyleProp<ViewStyle>;
    toggleButtonGroupCompButton?: StyleProp<ViewStyle>;
    toggleButtonGroupCompSelectedButton?: StyleProp<ViewStyle>;
    toggleButtonGroupCompButtonText?: StyleProp<TextStyle>;
    toggleButtonGroupCompButtonTextSelected?: StyleProp<TextStyle>;
  };
}

export const ToggleButtonGroupComp: React.FC<ToggleButtonGroupCompProps> = ({
  onValueChange,
  buttons,
  label,
  passStyles,
}) => {
  const styles = {
    toggleButtonGroupCompContainer:
      passStyles?.toggleButtonGroupCompContainer || {},
    toggleButtonGroupCompLabel: passStyles?.toggleButtonGroupCompLabel || {},
    toggleButtonGroupCompButtons:
      passStyles?.toggleButtonGroupCompButtons || {},
    toggleButtonGroupCompButton: passStyles?.toggleButtonGroupCompButton || {},
    toggleButtonGroupCompSelectedButton:
      passStyles?.toggleButtonGroupCompSelectedButton || {},
    toggleButtonGroupCompButtonText:
      passStyles?.toggleButtonGroupCompButtonText || {},
    toggleButtonGroupCompButtonTextSelected:
      passStyles?.toggleButtonGroupCompButtonTextSelected || {},
  };

  const [selectedValue, setSelectedValue] = useState("");

  const handleValueChange = (value: string) => {
    setSelectedValue(value);
    if (onValueChange) {
      onValueChange(value);
    }
  };

  const isSelected = (value: string) => selectedValue === value;

  return (
    <View style={styles.toggleButtonGroupCompContainer}>
      <Text style={styles.toggleButtonGroupCompLabel}>{label}</Text>
      <View style={styles.toggleButtonGroupCompButtons}>
        {buttons.map((buttonName) => (
          <Pressable
            key={buttonName}
            testID="button"
            style={[
              styles.toggleButtonGroupCompButton,
              isSelected(buttonName) &&
                styles.toggleButtonGroupCompSelectedButton,
            ]}
            onPress={() => handleValueChange(buttonName)}
          >
            <Text
              style={[
                styles.toggleButtonGroupCompButtonText,
                isSelected(buttonName) &&
                  styles.toggleButtonGroupCompButtonTextSelected,
              ]}
            >
              {buttonName}
            </Text>
          </Pressable>
        ))}
      </View>
    </View>
  );
};

export default ToggleButtonGroupComp;
