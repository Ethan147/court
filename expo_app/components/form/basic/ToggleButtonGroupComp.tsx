import React, { useState } from "react";
import {
  View,
  Text,
  TouchableOpacity,
  StyleSheet,
  useWindowDimensions,
  Platform,
} from "react-native";
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp,
} from "react-native-responsive-screen";
import colors from "../../../utils/colors";
import theme from "../../../utils/theme";

interface ToggleButtonGroupCompProps {
  onValueChange?: (value: string) => void;
}

export const ToggleButtonGroupComp: React.FC<ToggleButtonGroupCompProps> = ({
  onValueChange,
}) => {
  const windowDimensions = useWindowDimensions();

  const styles = StyleSheet.create({
    container: {
      ...Platform.select({
        web: {
          paddingVertical: windowDimensions.width * 0.01,
          paddingHorizontal: windowDimensions.width * 0.01,
          marginRight: windowDimensions.width * 0.01,
        },
        ios: {
          paddingVertical: wp("1%"),
          paddingHorizontal: wp("1%"),
          marginRight: wp("1%"),
        },
        android: {
          // todo
        },
      }),
      justifyContent: "center",
      alignItems: "center",
    },
    label: {
      ...Platform.select({
        web: {
          marginBottom: windowDimensions.width * 0.02,
        },
        ios: {
          paddingVertical: wp("1%"),
        },
        android: {
          // todo
        },
      }),
      fontSize: theme.font.size.small,
      color: colors.textSecondary,
    },
    buttons: {
      flexDirection: "row",
    },
    button: {
      ...Platform.select({
        web: {
          paddingVertical: windowDimensions.width * 0.01,
          paddingHorizontal: windowDimensions.width * 0.01,
          marginRight: windowDimensions.width * 0.01,
        },
        ios: {
          paddingVertical: wp("1%"),
          paddingHorizontal: wp("1%"),
          marginRight: wp("1%"),
        },
        android: {
          // todo
        },
      }),
      borderColor: colors.textSecondary,
      borderWidth: 0.5,
      borderRadius: 12,
      justifyContent: "center",
      alignItems: "center",
    },
    buttonText: {
      fontSize: theme.font.size.medium,
    },
    selected: {
      backgroundColor: colors.primary,
    },
  });

  const [gender, setGender] = useState("male");

  const handleValueChange = (value: string) => {
    setGender(value);
    if (onValueChange) {
      onValueChange(value);
    }
  };

  const isSelected = (value: string) => gender === value;

  return (
    <View style={styles.container}>
      <Text style={styles.label}>what's your gender?</Text>
      <View style={styles.buttons}>
        <TouchableOpacity
          style={[styles.button, isSelected("male") && styles.selected]}
          onPress={() => handleValueChange("male")}
        >
          <Text style={styles.buttonText}>male</Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={[styles.button, isSelected("female") && styles.selected]}
          onPress={() => handleValueChange("female")}
        >
          <Text style={styles.buttonText}>female</Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={[styles.button, isSelected("other") && styles.selected]}
          onPress={() => handleValueChange("other")}
        >
          <Text style={styles.buttonText}>other / prefer not to say</Text>
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default ToggleButtonGroupComp;
