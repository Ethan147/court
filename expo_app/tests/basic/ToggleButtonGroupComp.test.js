import { render, fireEvent, cleanup } from "@testing-library/react-native";
import React from "react";

import ToggleButtonGroupComp from "../../components/basic/ToggleButtonGroupComp";

afterEach(cleanup); // Unmounts React trees to clean up between tests

describe("<ToggleButtonGroupComp />", () => {
  it("renders without crashing", () => {
    render(<ToggleButtonGroupComp buttons={[]} label="Test Label" />);
  });

  it("renders zero buttons when zero buttons are passed in", () => {
    const buttons = [];
    const { queryAllByTestId } = render(
      <ToggleButtonGroupComp buttons={buttons} label="Test Label" />
    );

    const renderedButtons = queryAllByTestId("button");
    expect(renderedButtons).toHaveLength(0);
  });

  it("renders three buttons when three buttons are passed in", () => {
    const buttons = ["button 1", "button 2", "button 3"];
    const { getAllByTestId } = render(
      <ToggleButtonGroupComp buttons={buttons} label="Test Label" />
    );

    const renderedButtons = getAllByTestId("button");
    expect(renderedButtons).toHaveLength(buttons.length);
  });

  it("renders nine buttons when nine buttons are passed in", () => {
    const buttons = ["1", "2", "3", "4", "5", "6", "7", "8", "9"];
    const { getAllByTestId } = render(
      <ToggleButtonGroupComp buttons={buttons} label="Test Label" />
    );

    const renderedButtons = getAllByTestId("button");
    expect(renderedButtons).toHaveLength(buttons.length);
  });

  it("isSelected returns true only if one of the buttons is selected", () => {
    const buttons = ["button 1", "button 2", "button 3"];
    const testStyles = {
      toggleButtonGroupCompButtonTextSelected: {
        color: "blue",
      },
    };
    const { getByText } = render(
      <ToggleButtonGroupComp
        buttons={buttons}
        label="Test Label"
        passStyles={testStyles}
      />
    );
    fireEvent.press(getByText("button 2"));

    expect(getByText("button 1")).not.toHaveStyle(
      testStyles.toggleButtonGroupCompButtonTextSelected
    );
    expect(getByText("button 2")).toHaveStyle(
      testStyles.toggleButtonGroupCompButtonTextSelected
    );
    expect(getByText("button 3")).not.toHaveStyle(
      testStyles.toggleButtonGroupCompButtonTextSelected
    );
  });

  it("sets selectedValue when a button is pressed", () => {
    const buttons = ["button 1", "button 2", "button 3"];
    const onValueChange = jest.fn();
    const { getByText } = render(
      <ToggleButtonGroupComp
        buttons={buttons}
        label="Test Label"
        onValueChange={onValueChange}
      />
    );
    fireEvent.press(getByText("button 1"));
    expect(onValueChange).toHaveBeenCalledWith("button 1");
  });

  it("renders the provided label", () => {
    const buttons = ["button 1", "button 2", "button 3"];
    const label = "Test Label";
    const { getByText } = render(
      <ToggleButtonGroupComp buttons={buttons} label={label} />
    );
    expect(getByText(label)).toBeTruthy();
  });

  it("matches the snapshot", () => {
    const buttons = ["button 1", "button 2", "button 3"];
    const tree = render(
      <ToggleButtonGroupComp buttons={buttons} label="Test Label" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
