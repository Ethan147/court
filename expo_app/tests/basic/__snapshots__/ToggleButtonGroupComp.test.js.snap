// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleButtonGroupComp /> matches the snapshot 1`] = `
<View
  style={{}}
>
  <Text
    style={{}}
  >
    Test Label
  </Text>
  <View
    style={{}}
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {},
          false,
        ]
      }
      testID="button"
    >
      <Text
        style={
          [
            {},
            false,
          ]
        }
      >
        button 1
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {},
          false,
        ]
      }
      testID="button"
    >
      <Text
        style={
          [
            {},
            false,
          ]
        }
      >
        button 2
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {},
          false,
        ]
      }
      testID="button"
    >
      <Text
        style={
          [
            {},
            false,
          ]
        }
      >
        button 3
      </Text>
    </View>
  </View>
</View>
`;
